### PR TITLE
PRO-1272 Bugfix: Fix Broken Type Mapping

### PR DESCRIPTION
**Pull request by cedricwider**
## Jira

https://jira.localsearch.ch/browse/pro-1272

## Changes

- Add specs to prevent regressions and for faster bugfixing

### CODE REVIEW

File: spec/record/compact_spec.rb [hunk 1]
```diff
@@ -51,17 +66,25 @@ class User < LHS::Record
...
+      expect(places.compact.length).to eq 1
+      expect(places.length).not_to eq 1 # leaves the original intact
+    end
+  end
+
+  context 'record casting' do
+    let(:expected_display_name) { '*Mc Donalds*' }
+
+    it 'finds the right record class' do
+      expect(places.first.display_name).to eq expected_display_name
```
Comment (cedricwider) [inline spec/record/compact_spec.rb:79]: This passes as you'd expect.

File: spec/record/compact_spec.rb [hunk 2]
```diff
@@ -51,17 +66,25 @@ class User < LHS::Record
...
+      expect(places.map(&:display_name)).to eq [expected_display_name]
```
Comment (cedricwider) [inline spec/record/compact_spec.rb:80]: ...but this fails for whatever reason. :confused:

File: lib/lhs/concerns/collection/internal_collection.rb [hunk 3]
```diff
@@ -47,18 +47,26 @@ def compact!
...
+      end
+
+      def record_by_href(item)
+        return if json_value_type?(item) || (item[:href].blank? && @record.blank?)
+
+        LHS::Record.for_url(item[:href]) || @record
+      end
+
+      def json_value_type?(item)
+        item.is_a?(String) || item.is_a?(Numeric) || item.is_a?(TrueClass) || item.is_a?(FalseClass)
```
Comment (spape) [inline lib/lhs/concerns/collection/internal_collection.rb:69]: `json_value_type` seems misleading as neither String, Numeric, True or false have something to do with json (well string does). Would suggest something like `plain_value?`

File: lib/lhs/concerns/collection/internal_collection.rb [hunk 4]
```diff
@@ -47,18 +47,26 @@ def compact!
...
       def cast_item(item)
-        record_by_href = LHS::Record.for_url(item[:href]) if item[:href]
         data = LHS::Data.new(item, @parent, @record)
-        return (record_by_href || @record).new(data) if record_by_href || @record
-        data
+        record_by_href(item)&.new(data) || data
+      end
+
+      def record_by_href(item)
+        return if json_value_type?(item) || (item[:href].blank? && @record.blank?)
```
Comment (spape) [inline lib/lhs/concerns/collection/internal_collection.rb:63]: there should be no `@record` in record_by_href. as `record_by_href` implies that it provides, well a record by href not by instance.

Comment (spape) [reply lib/lhs/concerns/collection/internal_collection.rb]: maybe remove it from `record_by_href` and try something like (in `cast_item`):

```ruby
(record_by_href(item) || @record)&.new(data) || data
```

**Review by spape — APPROVED** [review]
gracias

**Pull request merged**